File: storage/scripts/leadSNP.R
library(data.table)
args <- commandArgs(TRUE)
filedir <- args[1]
snps <- fread(paste(filedir, "snps.txt", sep=""), data.table=F)
ld <- fread(paste(filedir, "ld.txt", sep=""), data.table=F)
r2th <- as.numeric(args[2])
gwasPth <- as.numeric(args[3])
leadP <- as.numeric(args[4])
minMAF <- as.numeric(args[5])
mergeDist <- as.numeric(args[6])*1000
inlead <- args[7]
if(inlead != 0){
  inlead <- fread(paste(filedir, "input.lead", sep=""), data.table=F)
}else{
  inlead <- data.frame(matrix(vector(), 0, 3))
}
colnames(inlead) <- c("rsID", "chr", "pos")
inlead$uniqID <- snps$uniqID[match(inlead$rsID, snps$rsID)]


ld <- ld[ld$r2>=r2th,]
snps <- snps[(snps$gwasP<=gwasPth | is.na(snps$gwasP) | snps$uniqID %in% inlead$uniqID) & snps$MAF>=minMAF & snps$uniqID %in% ld$SNP2,]
ld <- ld[ld$SNP1 %in% snps$uniqID[snps$gwasP <= leadP | snps$uniqID %in% inlead$uniqID],]
ld <- ld[ld$SNP2 %in% snps$uniqID,]

inlead <- inlead[inlead$uniqID %in% snps$uniqID,]

indS <- data.frame(matrix(vector(), 0, 8, dimnames = list(c(),c("No","uniqID", "rsID", "chr", "pos", "p", "nSNPs", "nGWASSNPs"))))
leadSNPs <- unique(ld$SNP1)

while(length(leadSNPs)>0){
  j <- nrow(indS)+1
  ls <- snps$uniqID[snps$uniqID %in% leadSNPs][which.min(snps$gwasP[snps$uniqID %in% leadSNPs])]
  indS[j,2] <- ls
  indS[j,3] <- snps$rsID[which(snps$uniqID==ls)]
  indS[j,4] <- snps$chr[which(snps$uniqID==ls)]
  indS[j,5] <- snps$pos[which(snps$uniqID==ls)]
  indS[j,6] <- snps$gwasP[which(snps$uniqID==ls)]
  indS[j,7] <- length(which(ld$SNP1==ls))
  indS[j,8] <- length(which(ld$SNP2[ld$SNP1==ls] %in% snps$uniqID[!is.na(snps$gwasP)]))

  leadSNPs <- leadSNPs[!(leadSNPs %in% ld$SNP2[which(ld$SNP1==ls)])]
}

indS <- indS[order(indS$pos),]
indS <- indS[order(indS$chr),]
indS$No <- 1:nrow(indS)

ld <- ld[ld$SNP1 %in% indS$uniqID,]
snps <- snps[snps$uniqID %in% ld$SNP2,]

indS$GenomicLocus <- NA
loci <- data.frame(matrix(vector(), 0, 12, dimnames=list(c(), c("GenomicLocus","uniqID", "rsID", "chr", "pos", "p", "start", "end", "nIndSigSNPs", "IndSigSNPs", "nSNPs", "nGWASSNPs"))))
j<-1
loci[j,1] <- j
loci[j,2] <- indS$uniqID[1]
loci[j,3] <- indS$rsID[1]
loci[j,4] <- indS$chr[1]
loci[j,5] <- indS$pos[1]
loci[j,6] <- indS$p[1]
loci[j,9] <- 1
loci[j,7] <- min(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[j]]])
loci[j,8] <- max(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[j]]])
loci[j,10] <- indS$rsID[1]
indS$GenomicLocus[j] <- 1
if(nrow(indS)>1){
  for(i in 2:nrow(indS)){
    if(loci$chr[j]==indS$chr[i] & min(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[i]]])-loci$end[j]<=mergeDist){
      if(indS$p[i] < loci[j,6]){
      loci[j,2] <- indS$uniqID[i]
      loci[j,3] <- indS$rsID[i]
      loci[j,4] <- indS$chr[i]
      loci[j,5] <- indS$pos[i]
      loci[j,6] <- indS$p[i]
      }
      loci[j,9] <- loci[j,7]+1
      loci[j,7] <- min(c(loci[j,8], min(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[i]]])))
      loci[j,8] <- max(c(loci[j,9], max(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[i]]])))
      loci[j,10] <- paste(loci[j,10], indS$rsID[i], sep=":")
      indS$GenomicLocus[i] <- j
    }else{
      j <- j+1
      loci[j,1] <- j
      loci[j,2] <- indS$uniqID[i]
      loci[j,3] <- indS$rsID[i]
      loci[j,4] <- indS$chr[i]
      loci[j,5] <- indS$pos[i]
      loci[j,6] <- indS$p[i]
      loci[j,9] <- 1
      loci[j,7] <- min(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[i]]])
      loci[j,8] <- max(snps$pos[snps$uniqID %in% ld$SNP2[ld$SNP1==indS$uniqID[i]]])
      loci[j,10] <- indS$rsID[i]
      indS$GenomicLocus[i] <- j
    }
  }
}

loci <- loci[,c(1:8,11,12,9,10)]

snps$r2 <- sapply(snps$uniqID, function(x){max(ld$r2[ld$SNP2==x])})
snps$IndSigSNP <- sapply(snps$uniqID, function(x){paste(snps$rsID[snps$uniqID %in%ld$SNP1[ld$SNP2==x & ld$r2==max(ld$r2[ld$SNP2==x])]], collapse=":")})
snps$GenomicLocus <- NA
for(i in 1:nrow(loci)){
  ls <- snps$uniqID[snps$rsID %in% unlist(strsplit(loci$IndSigSNPs[i], ":"))]
  loci$nSNPs[i] <- length(unique(ld$SNP2[ld$SNP1 %in% ls]))
  loci$nGWASSNPs[i] <- length(unique(ld$SNP2[ld$SNP1 %in% ls & ld$SNP2 %in% snps$uniqID[!is.na(snps$gwasP)]]))
  snps$GenomicLocus[snps$chr==loci$chr[i]&snps$pos>=loci$start[i] & snps$pos<=loci$end[i]] <- i
}

annov.input <- data.frame(chr=snps$chr, start=snps$pos, end=snps$pos, ref=snps$ref, alt=snps$alt)
#n <- t(sapply(snps$uniqID, function(x){unlist(strsplit(x, ":"))[3:4]}))
#annov.input$ref <- n[,1]
#annov.input$alt <- n[,2]

indS <- subset(indS, select=c("No","GenomicLocus", "uniqID", "rsID", "chr", "pos", "p", "nSNPs", "nGWASSNPs"))

write.table(indS, paste(filedir,"IndSigSNPs.txt",sep=""), quote=F, row.names=F, sep="\t")
write.table(loci, paste(filedir, "GenomicRiskLoci.txt", sep=""), quote=F, row.names=F, sep="\t")
write.table(snps, paste(filedir,"snps.txt",sep=""), quote=F, row.names=F, sep="\t")
write.table(ld, paste(filedir,"ld.txt",sep=""), quote=F, row.names=F, sep="\t")
write.table(annov.input, paste(filedir, "annov.input", sep=""), quote=F, row.names=F, col.names=F, sep="\t")

File: storage/scripts/leadSNP.py
#!/usr/bin/python

import sys
import os
import re
import pandas as pd
import numpy as np
import tabix
import ConfigParser

##### Return index of a1 which exists in a2 #####
def ArrayIn(a1, a2):
	# results = [i for i, x in enumerate(a1) if x in a2]
	results = np.where(np.in1d(a1, a2))[0]
	return results


filedir = sys.argv[1]
if re.match(".+\/$", filedir) is None:
	filedir += '/'

cfg = ConfigParser.ConfigParser()
cfg.read(os.path.dirname(os.path.realpath(__file__))+'/app.config')

param = ConfigParser.ConfigParser()
param.read(filedir+'params.config')

ldfiledir = cfg.get('data', 'refgenome')
pop = param.get('params', 'pop')
# r2 = param.get('params', 'r2')
# gwasP = param.get('params', 'gwasP')
# leadP = param.get('params', 'leadP')
# maf = param.get('params', 'MAF')
# mergeDist = param.get('params', 'mergeDist')
# inlead = param.get('inputfiles', 'leadSNPsfile')
# if inlead=="NA":
# 	inlead = None
#
# snps = pd.read_table(filedir+"snps.txt")
# snpshead = snps.columns.values
# snps = snps.as_matrix()
# ld = pd.read_table(filedir+"ld.txt")
# ld = ld.as_matrix()
#
# ld = ld[ld[:,2]>=r2]


indSNPs = pd.read_table(filedir+'IndSigSNPs.txt')
indSNPs = indSNPs.as_matrix()
indSNPs = indSNPs[indSNPs[:,6].astype(float).argsort()]

checked = []
leadSNPs = []
# leadSNPs.append(["No", "GenomicLocus", "uniqID", "chr", "pos", "rsID", "P", "IndSigSNPs"])

for snp in indSNPs:
    if snp[3] in checked:
        continue
    ldfile = ldfiledir+'/'+pop+'/'+pop+'.chr'+str(snp[4])+'.ld.gz';
    tb = tabix.open(ldfile)
    ld = tb.querys(str(snp[4])+":"+str(snp[5])+"-"+str(snp[5]))
    inSNPs = []
    inSNPs.append(snp[3])
    for l in ld:
        if float(l[4])<0.1:
            continue
        if l[3] in indSNPs[:,3]:
            checked.append(l[3])
            inSNPs.append(l[3])
    leadSNPs.append([len(leadSNPs)+1, snp[1], snp[2], snp[4], snp[5], snp[3], snp[6], len(inSNPs), ":".join(inSNPs)])

leadSNPs = np.array(leadSNPs)
# leadSNPs = leadSNPs[leadSNPs[:,3].astype(int).argsort()]
# leadSNPs = leadSNPs[leadSNPs[:,2].astype(int).argsort()]
leadSNPs = leadSNPs[np.lexsort((leadSNPs[:,4].astype(int),leadSNPs[:,3].astype(int)))]

for i in range(0, len(leadSNPs)):
    leadSNPs[i,0] = i+1

fout = open(filedir+"leadSNPs.txt", 'w')
fout.write("\t".join(["No", "GenomicLocus", "uniqID", "chr", "pos", "rsID", "p", "nIndSigSNPs", "IndSigSNPs"])+"\n")
for l in leadSNPs:
    fout.write("\t".join(l.astype(str))+"\n")
fout.close()

loci = pd.read_table(filedir+"GenomicRiskLoci.txt")
locihead = list(loci.columns.values)
locihead.append("nLeadSNPs")
locihead.append("LeadSNPs")

loci = np.c_[loci, [0]*len(loci), [0]*len(loci)]
for i in range(0, len(loci)):
	loci[i,12] = sum(x==loci[i,0] for x in leadSNPs[:,1].astype(int))
	loci[i,13] = ":".join(leadSNPs[leadSNPs[:,1].astype(int)==loci[i,0],5])

fout = open(filedir+"GenomicRiskLoci.txt", 'w')
fout.write("\t".join(locihead)+"\n")
for l in loci:
	fout.write("\t".join(l.astype(str))+"\n")
fout.close()
